README.md
# perfSONAR Lookup Service Cache Daemon

The Lookup Service Cache Daemon periodically downloads a tarball of CSV files using a conditional HTTP GET containing a listing of perfSONAR Services. This code does NOT generate the CSV files, it simply handles downloading them periodically. These files are used to then populate the Toolkit GUI's test selection interface.

##Getting the Code
You may checkout the code with the following command:

```
git clone --recursive https://github.com/perfsonar/ls-cache-daemon.git
```

Note the use of the `--recursive` option to ensure any submodule trees are included in the clone.

##Building and Installing

To install the code on your system run:

```bash
make install
```

##Packaging
You may create a source tarball of this code with the following:

```bash
make dist
```
##Running 

To start the service run:

```bash
/etc/init.d/perfsonar-lscachedaemon start
```

To stop the service run:

```bash
/etc/init.d/perfsonar-lscachedaemon stop
```

To restart the service run:

```bash
/etc/init.d/perfsonar-lscachedaemon restart
```

##Using the *shared* Submodule
This repository contains a [git submodule](http://git-scm.com/book/en/v2/Git-Tools-Submodules) to the perfSONAR [shared](https://github.com/perfsonar/perl-shared) repository. This submodule is used to access common perfSONAR libraries. You will find a number of symbolic links to these modules under *lib*. The use of a submodule has a few implications when working with the code in this repository:

* As previously noted, when you clone the repository for the first time, you will want to use the `--recursive` option to make sure the submodule tree is included. If you do not, any symbolic links under *lib* will be broken in your local copy. If you forget the `--recursive` option, you can pull the submodule tree with the following commands:

    ```bash
    git submodule init
    git submodule update
    ```
* When you are editing files under *lib* be sure to check if you are working on an actual file or whether it's a link to something under *shared*. In general it is better to make changes to the *shared* submodule by editing the *shared* repository directly. If however you do make changes while working in this repository, see the [git submodule page](http://git-scm.com/book/en/v2/Git-Tools-Submodules#Working-on-a-Project-with-Submodules) for more details on pushing those changes to the server.
* Keep in mind that a submodule points at a specific revision of the repository it is referencing. As such if a new commit is made to the shared submodule's repository, this project will not get the change automatically. Instead it will still point at the old revision. To update to the latest revision of the *shared* submodule repository run the following commands:

    ```bash
    git submodule foreach git pull origin master
    git commit -a -m "Updating to latest shared"
    git push
    ```
* If you want to include a new file from the *shared* submodule, create a symbolic link under *lib*. For example, if you were to add a reference to the  *perfSONAR_PS::Utils::DNS* module you would run the following:

    ```bash
    mkdir -p lib/perfSONAR_PS/Utils/
    cd lib/perfSONAR_PS/Utils/
    ln -s ../../../shared/lib/perfSONAR_PS/Utils/DNS.pm DNS.pm
    ```
For more information on using the submodule, see the *shared/README.md* file or access it [here](https://github.com/perfsonar/perl-shared/blob/master/README.md) 



debian/changelog
perfsonar-lscachedaemon (4.4.6-1) perfsonar-4.4; urgency=low

  * New upstream version.

 -- perfSONAR developers <debian@perfsonar.net>  Mon, 07 Nov 2022 11:00:52 -0500

perfsonar-lscachedaemon (4.4.5-1) perfsonar-4.4; urgency=low

  * New upstream version.

 -- perfSONAR developers <debian@perfsonar.net>  Tue, 20 Sep 2022 08:42:44 -0400

perfsonar-lscachedaemon (4.4.4-1) perfsonar-4.4; urgency=low

  * New upstream version.

 -- perfSONAR developers <debian@perfsonar.net>  Mon, 04 Apr 2022 12:17:11 -0400

perfsonar-lscachedaemon (4.4.3-1) perfsonar-4.4; urgency=low

  * New upstream version.

 -- perfSONAR developers <debian@perfsonar.net>  Thu, 03 Feb 2022 15:55:43 -0500

perfsonar-lscachedaemon (4.4.2-1) perfsonar-4.4; urgency=low

  * New upstream version.

 -- perfSONAR developers <debian@perfsonar.net>  Wed, 12 Jan 2022 09:33:53 -0500

perfsonar-lscachedaemon (4.4.1-1) perfsonar-4.4; urgency=low

  * New upstream version.

 -- Antoine Delvaux <antoine.delvaux@man.poznan.pl>  Thu, 26 Aug 2021 01:31:10 +0100

perfsonar-lscachedaemon (4.4.0-1) perfsonar-4.4; urgency=low

  * New upstream version.
  * debhelper 9.20160709 provides dh-systemd

 -- perfSONAR developers <debian@perfsonar.net>  Wed, 07 Jul 2021 15:20:16 +0000

perfsonar-lscachedaemon (4.3.4-1) perfsonar-4.3; urgency=low

  * New upstream version.

 -- perfSONAR developers <debian@perfsonar.net>  Wed, 17 Mar 2021 01:10:23 +0000

perfsonar-lscachedaemon (4.3.3-1) perfsonar-4.3; urgency=low

  * New upstream version.

 -- perfSONAR developers <debian@perfsonar.net>  Tue, 02 Feb 2021 16:37:45 +0000

perfsonar-lscachedaemon (4.3.2-1) perfsonar-4.3; urgency=low

  * New upstream version.

 -- perfSONAR developers <debian@perfsonar.net>  Wed, 18 Nov 2020 15:43:28 +0000

perfsonar-lscachedaemon (4.3.0-1) perfsonar-4.3; urgency=low

  * New upstream version.

 -- perfSONAR developers <debian@perfsonar.net>  Fri, 30 Oct 2020 13:51:27 +0000

perfsonar-lscachedaemon (4.2.4-1) perfsonar-4.2; urgency=low

  * New upstream version.

 -- perfSONAR developers <debian@perfsonar.net>  Thu, 26 Mar 2020 10:03:47 -0400

perfsonar-lscachedaemon (4.2.3-1) perfsonar-4.2; urgency=low

  * New upstream version.

 -- perfSONAR developers <debian@perfsonar.net>  Wed, 19 Feb 2020 15:38:16 +0000

perfsonar-lscachedaemon (4.2.1-1) perfsonar-4.2; urgency=low

  * New upstream version.

 -- perfSONAR developers <debian@perfsonar.net>  Fri, 27 Sep 2019 13:17:18 -0400

perfsonar-lscachedaemon (4.2.0-1) perfsonar-minor-staging; urgency=low

  * New upstream version
  * Building on Stretch.

 -- Andy Lake <andy@es.net>  Thu, 15 Aug 2019 01:11:00 +0000

perfsonar-lscachedaemon (4.1.3-1) perfsonar-release; urgency=low

  * New upstream version

 -- Antoine Delvaux <antoine.delvaux@man.poznan.pl>  Sat, 13 Oct 2018 21:34:12 +0000

perfsonar-lscachedaemon (4.1-1) perfsonar-release; urgency=low

  * New upstream version

 -- Antoine Delvaux <antoine.delvaux@man.poznan.pl>  Mon, 13 Aug 2018 18:48:08 +0000

perfsonar-lscachedaemon (4.0.2-1) perfsonar-jessie-release; urgency=low

  * New upstream version

 -- Antoine Delvaux <antoine.delvaux@man.poznan.pl>  Tue, 19 Dec 2017 16:14:50 +0000

perfsonar-lscachedaemon (4.0-1) perfsonar-jessie-release; urgency=low

  * New upstream version
  * Adding autopkgtests

 -- Antoine Delvaux <antoine.delvaux@man.poznan.pl>  Thu, 13 Apr 2017 16:37:27 +0000

perfsonar-lscachedaemon (3.5.1-2) perfsonar-jessie-release; urgency=medium

  * Add liblog-dispatch-filerotate-perl dependency

 -- Valentin Vidic <Valentin.Vidic@CARNet.hr>  Tue, 15 Mar 2016 15:18:22 +0100

perfsonar-lscachedaemon (3.5.1-1) perfsonar-jessie-release; urgency=medium

  * Initial Debian packaging.

 -- Valentin Vidic <Valentin.Vidic@CARNet.hr>  Fri, 04 Mar 2016 10:44:52 +0100


lib/perfSONAR_PS/Utils/NetLogger.pm
package perfSONAR_PS::Utils::NetLogger;

use strict;
use warnings;

our $VERSION = 3.3;

=head1 NAME
     
perfSONAR_PS::Utils::NetLogger
    
=head1 DESCRIPTION

A module that provides tools to generate NetLogger formatted messages for
log4perl.  For more information on NetLogger see: 

http://dsd.lbl.gov/NetLoggerWiki/index.php/Main_Page

=head1 API
    
The API of NetLogger is used to format log messages in the NetLogger 'Best
Practices' format. See: http://www.cedps.net/wiki/index.php/LoggingBestPractices
    
=cut

require 5.002;
use Time::HiRes;
use Data::UUID;

# initialize Global GUID
my $GUID = get_guid();

=head2 format("event_name", list of name=>value pairs)

Sample use:

  use Log::Log4perl qw(:easy);
  use NetLogger;
  Log::Log4perl->easy_init($DEBUG);

  my $logger = get_logger("my_prog");

  $logger->info(NetLogger::format("org.perfsonar.client.parseResults.start"));
  # call function here
  $logger->info(NetLogger::format("org.perfsonar.client.parseResults.end", {val=>12,}));
  

This will generate a log that looks like this:

2007/12/19 13:51:26 39899 INFO> myprog:NN main:: - ts=2007-12-19T21:51:26.030823Z \
	event=org.perfsonar.client.runQuery1.end guid=736ee764-ae7c-11dc-9f7d-000f1f6ed15d

=cut

sub format {
    my ( $evnt, $data ) = @_;
    my ( $str ) = q{};
    if ( exists $data->{'ts'} ) {
        $str = "ts=$data->{ 'ts' } ";
    }
    else {
        my $dt = date();
        $str = "ts=$dt ";
    }
    $str .= "event=$evnt ";
    foreach my $k ( keys %$data ) {
        $str .= "$k=$data->{$k} ";
    }
    $str .= "guid=" . $GUID;
    return $str;
}

=head2 date

TBD

=cut

sub date {
    my ( $tm, $usec ) = Time::HiRes::gettimeofday();
    my ( $sec, $min, $hour, $mday, $mon, $year, $wday, $yday, $isdst ) = gmtime( $tm );
    return sprintf( "%04d-%02d-%02dT%02d:%02d:%02d.%06dZ", $year + 1900, $mon + 1, $mday, $hour, $min, $sec, $usec );
}

=head2 get_guid

TBD

=cut

sub get_guid {
    my $ug   = new Data::UUID;
    my $guid = $ug->create_str();

    # if dont have Data::UUID, can do this instead
    #  my $guid = `uuidgen`; chomp $guid;

    return ( $guid );
}

=head2 reset_guid

TBD

=cut

sub reset_guid {
    my $ug = new Data::UUID;
    $GUID = $ug->create_str();

    # if dont have Data::UUID, can do this instead
    #  my $GUID = `uuidgen`; chomp $GUID;

    return;
}

1;

__END__

=head1 SEE ALSO

L<Time::HiRes>, L<Data::UUID>

To join the 'perfSONAR-PS Users' mailing list, please visit:

  https://lists.internet2.edu/sympa/info/perfsonar-ps-users

The perfSONAR-PS git repository is located at:

  https://code.google.com/p/perfsonar-ps/

Questions and comments can be directed to the author, or the mailing list.
Bugs, feature requests, and improvements can be directed here:

  http://code.google.com/p/perfsonar-ps/issues/list

=head1 VERSION

$Id: NetLogger.pm 2640 2009-03-20 01:21:21Z zurawski $

=head1 AUTHOR

Dan Gunter, dkgunter@lbl.gov

=head1 LICENSE

See: http://dsd.lbl.gov/NetLoggerWiki/index.php/Licensing

=head1 COPYRIGHT

Copyright (c) 2004-2007, Internet2, Lawrenence Berkeley National Lab, and the
University of California.

All rights reserved.

=cut
